### PR TITLE
Add the agreement exceptions text to the returns requirements check page

### PR DESCRIPTION
**Pull request by jonathangoulding**
Due to the estimated complexity of adding fetch services for purposes, points and other display items. We have broken down the complex display items into their own branches of work.

https://eaflood.atlassian.net/browse/WATER-4386

This branch will add agreement exceptions to the return requirement summary card. This will be a comma seperated list in the oxford comma style.

Previous work done here - #1019

### CODE REVIEW

File: app/presenters/return-requirements/check/returns-requirements.presenter.js [hunk 1]
```diff
@@ -39,6 +47,32 @@ function _abstractionPeriod (abstractionPeriod) {
...
+    return agreementsExceptionsText[exception]
+  })
+
+  if (formattedExceptions.length === 1) {
+    return formattedExceptions[0]
+  } else if (formattedExceptions.length === 2) {
+    return formattedExceptions.join(' and ')
+  } else {
+    return _formatExceptionsToSerialCommaList(formattedExceptions)
+  }
```
Comment (Cruikshanks) [inline app/presenters/return-requirements/check/returns-requirements.presenter.js:65]: Me and my guard clauses 😁 

```suggestion
  if (formattedExceptions.length === 1) {
    return formattedExceptions[0]
  }
    
  if (formattedExceptions.length === 2) {
    return formattedExceptions.join(' and ')
  }
    
  return _formatExceptionsToSerialCommaList(formattedExceptions)
```

File: app/presenters/return-requirements/check/returns-requirements.presenter.js [hunk 2]
```diff
@@ -39,6 +47,32 @@ function _abstractionPeriod (abstractionPeriod) {
...
+}
+
+function _formatExceptionsToSerialCommaList (exceptions) {
+  return exceptions
+    .slice(0, exceptions.length - 1)
+    .join(', ') +
+    (exceptions.length > 2 ? ',' : '') +
+    ' and ' + exceptions[exceptions.length - 1]
+}
+
```
Comment (Cruikshanks) [inline app/presenters/return-requirements/check/returns-requirements.presenter.js:75]: We're only calling this function if there are more than 2 `agreementExceptions`. Which means I think we can ditch the `exceptions.length > 2` check.

That makes things simple enough I'd have no reservations about moving this

```javascript
  return formattedExceptions
    .slice(0, formattedExceptions.length - 1)
    .join(', ') +
    ' and ' + formattedExceptions[formattedExceptions.length - 1]
```

into `_agreementsExceptions()` and dropping the function altogether.

File: test/presenters/return-requirements/check/returns-requirements.presenter.test.js [hunk 3]
```diff
@@ -141,6 +143,50 @@ describe('Return Requirements presenter', () => {
           })
         })
       })
+
+      describe('and the agreement exceptions', () => {
+        describe('is \'none\'', () => {
```
Comment (Cruikshanks) [inline test/presenters/return-requirements/check/returns-requirements.presenter.test.js:148]: NIT - [Our new convention](https://github.com/DEFRA/water-abstraction-team/blob/main/coding_conventions.md#unit-test-descriptive-text)

```suggestion
        describe('is "none"', () => {
```

File: test/presenters/return-requirements/check/returns-requirements.presenter.test.js [hunk 4]
```diff
@@ -141,6 +143,50 @@ describe('Return Requirements presenter', () => {
...
+          beforeEach(() => {
+            requirements = [{ ...requirement, agreementsExceptions: ['none'] }]
+          })
+
+          it('should return \'None\' ', () => {
```
Comment (Cruikshanks) [inline test/presenters/return-requirements/check/returns-requirements.presenter.test.js:153]:
```suggestion
          it('should return "None"', () => {
```

File: test/presenters/return-requirements/check/returns-requirements.presenter.test.js [hunk 5]
```diff
@@ -141,6 +143,50 @@ describe('Return Requirements presenter', () => {
...
+
+            expect(result.requirements[0].agreementsExceptions).to.equal('Gravity fill')
+          })
+        })
+
+        describe('has two exceptions', () => {
+          beforeEach(() => {
+            requirements = [{ ...requirement, agreementsExceptions: ['gravity-fill', 'transfer-re-abstraction-scheme'] }]
+          })
+          it('should return the exceptions in the string equivalent seperated with an and \'exception1 and exception2\' ', () => {
```
Comment (Cruikshanks) [inline test/presenters/return-requirements/check/returns-requirements.presenter.test.js:172]:
```suggestion
          })

          it('should return the exceptions as "Gravity fill and Transfer re-abstraction scheme"', () => {
```

File: test/presenters/return-requirements/check/returns-requirements.presenter.test.js [hunk 6]
```diff
@@ -141,6 +143,50 @@ describe('Return Requirements presenter', () => {
...
+
+          it('should return \'None\' ', () => {
+            const result = ReturnRequirementsPresenter.go(requirements, purposes, points, journey)
+
+            expect(result.requirements[0].agreementsExceptions).to.equal('None')
+          })
+        })
+
+        describe('has one exception', () => {
+          it('should return the exception in the string equivalent', () => {
```
Comment (Cruikshanks) [inline test/presenters/return-requirements/check/returns-requirements.presenter.test.js:161]:
```suggestion
          it('should return the exception as "Gravity fill"', () => {
```

File: test/presenters/return-requirements/check/returns-requirements.presenter.test.js [hunk 7]
```diff
@@ -141,6 +143,50 @@ describe('Return Requirements presenter', () => {
...
+
+            expect(result.requirements[0].agreementsExceptions).to.equal('Gravity fill and Transfer re-abstraction scheme')
+          })
+        })
+
+        describe('has more than two exceptions', () => {
+          beforeEach(() => {
+            requirements = [{ ...requirement, agreementsExceptions: ['gravity-fill', 'transfer-re-abstraction-scheme', 'two-part-tariff', '56-returns-exception'] }]
+          })
+          it('should return the exceptions formatted in the string equivalent, in the serial comma style e.g \'exception1, exception2, and exception3\' ', () => {
```
Comment (Cruikshanks) [inline test/presenters/return-requirements/check/returns-requirements.presenter.test.js:183]:
```suggestion
          })

          it('should return the exceptions as "Gravity fill, Transfer re-abstraction scheme, Two-part tariff, and 56 returns exception"', () => {
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]

File: app/presenters/return-requirements/check/returns-requirements.presenter.js (hunk 2 above)
Comment (jonathangoulding) [reply app/presenters/return-requirements/check/returns-requirements.presenter.js:77]: `(exceptions.length > 2 ? ',' : '')` this adds the serial comma at the end of the join so there is a comma before the and.

Comment (jonathangoulding) [reply app/presenters/return-requirements/check/returns-requirements.presenter.js:77]: i can remove condition. and just have a + ','

Comment (jonathangoulding) [reply app/presenters/return-requirements/check/returns-requirements.presenter.js:77]: I'll just add the comma in the and condition. No idea why i didn't do that in the first placee

```
exceptions.slice(0, exceptions.length - 1).join(', ') + ', and ' + exceptions[exceptions.length - 1]
```

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**